### PR TITLE
Evaluate Boolean Expressions

### PR DESCRIPTION
**Pull request by RimjhimGupta**
This program is for the evaluation of the Boolean expression where a string consists of only 0, 1, A, B, C where
A = AND
B = OR
C = XOR
The evaluation is done from left to right. The length of string will be odd. It will always be a valid string.
This pull request is for Hacktober Fest 2020.

# Welcome to Dart community

### **Describe your change:**

* [ ] Add an algorithm?
* [ ] Fix a bug or typo in an existing algorithm?
* [ ] Documentation change?

### **Checklist:**
* [ ] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Dart/blob/master/CONTRIBUTING.md).
* [ ] This pull request is all my own work -- I have not plagiarized.
* [ ] I know that pull requests will not be merged if they fail the automated tests.
* [ ] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [ ] All new Dart files are placed inside an existing directory.
* [ ] All new algorithms have a URL in its comments that points to Wikipedia or other similar explanation.
* [ ] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

File: maths/Boolean_expression.dart [hunk 1]
```diff
@@ -0,0 +1,40 @@
+String evaluateBoolExpr(String s) {
```
Comment (Parowicz) [inline maths/Boolean_expression.dart:1]: You could add some tests
```suggestion
import 'package:test/test.dart';

String evaluateBoolExpr(String s) {
```

File: maths/Boolean_expression.dart [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+        s.replaceRange(i + 2, i + 3, '1');
+    }
+  }
+  return s.substring(n - 1, n);
+}
+
+// Driver code
+void main() {
+  String s = "1A0B1";
+  print(evaluateBoolExpr(s));
```
Comment (Parowicz) [inline maths/Boolean_expression.dart:39]: Actually they do fail, could you check it?
```suggestion
  test("Test singleton boolean remains same", () {
    expect(evaluateBoolExpr("1"), equals('1'));
    expect(evaluateBoolExpr("0"), equals('0'));
  });

  test("Test AND", () {
    expect(evaluateBoolExpr("1A1"), equals('1'));
    expect(evaluateBoolExpr("0A1"), equals('0'));
    expect(evaluateBoolExpr("1A0"), equals('0'));
    expect(evaluateBoolExpr("0A0"), equals('0'));
  });

  test("Test OR", () {
    expect(evaluateBoolExpr("1B1"), equals('1'));
    expect(evaluateBoolExpr("0B1"), equals('1'));
    expect(evaluateBoolExpr("1B0"), equals('1'));
    expect(evaluateBoolExpr("0B0"), equals('0'));
  });

  test("Test XOR", () {
    expect(evaluateBoolExpr("1C1"), equals('0'));
    expect(evaluateBoolExpr("0C1"), equals('1'));
    expect(evaluateBoolExpr("1C0"), equals('1'));
    expect(evaluateBoolExpr("0C0"), equals('0'));
  });

  test("Test '0B0B0B0B0B1'", () {
    expect(evaluateBoolExpr("0B0B0B0B0B1"), equals('1'));
  });

  test("Test '0B0B0B1A1'", () {
    expect(evaluateBoolExpr("0B0B0B1A1"), equals('1'));
  });
```

**Review by Parowicz — CHANGES_REQUESTED** [review]

**Comment by RimjhimGupta** [comment]
@Parowicz Sir, I request you to please help me in resolving the failed check.

**Comment by Parowicz** [comment]
Tests are failing because  
1. `evaluateBoolExpr("0A1")` returns `'1'`   ==>  `false` and `true`  != `true`
2. `evaluateBoolExpr("1B0")` returns `'0'`   ==>  `true` or `false`     != `false` 
3. `evaluateBoolExpr("1C0")` returns `'0'`   ==>  `true` xor `false`   != `false`

**Comment by RimjhimGupta** [comment]
@Parowicz Sir, regarding the failing of tests, I had already specified in the description that evaluation is done from left to right only. But, you are trying to evaluate it from right to left.
Please consider this Sir.
Thank you.

**Comment by Parowicz** [comment]
Those operations are commutative https://en.wikipedia.org/wiki/Boolean_algebra#Laws
In your code `evaluateBoolExpr("1A0")` is different from `evaluateBoolExpr("0A1")`.
Evaluation strategy should not invalidate result.

**Pull request closed without merging**